packages/terra-button/src/Button.jsx

```diff
@@ -6,105 +6,238 @@ import styles from './Button.scss';
+   * Callback function triggered when clicked.
    */
   onClick: PropTypes.func,
   /**
-   * Sets the button size. One of `tiny`, `small`, `medium`, `large`, `huge`
+   * Callback function triggered when button loses focus.
    */
-  size: PropTypes.oneOf(['tiny', 'small', 'medium', 'large', 'huge']),
+  onBlur: PropTypes.func,
   /**
-   * Sets the button text
+   * Callback function triggered when button gains focus.
    */
-  text: PropTypes.string,
+  onFocus: PropTypes.func,
   /**
-   * Sets the button type. One of `button`, `submit`, or `reset`
+   * Callback function triggered when key is pressed.
+   */
+  onKeyDown: PropTypes.func,
+  /**
+   * Callback function triggered when key is released.
+   */
+  onKeyUp: PropTypes.func,
+  /**
+   *  React `ref` callback to receive and store a reference to the mounted button instance.
+   */
+  refCallback: PropTypes.func,
+  /**
+   * Sets the button text.
+   */
+  text: PropTypes.string.isRequired,
+  /**
+   * Sets the button type. One of `button`, `submit`, or `reset`.
    */
   type: PropTypes.oneOf(['button', 'submit', 'reset']),
   /**
-   * Sets the button variant. One of `primary`, `secondary`, or `link`
+   * Sets the button variant. One of `neutral`,  `emphasis`, `de-emphasis`, `utility` or `action`.
    */
-  variant: PropTypes.oneOf(['default', 'link', 'primary', 'secondary']),
+  variant: PropTypes.oneOf(['neutral', 'emphasis', 'de-emphasis', 'utility', 'action']),
 };
 
 const defaultProps = {
   isBlock: false,
-  isCompact: false,
   isDisabled: false,
+  isIconOnly: false,
   isReversed: false,
+  refCallback: undefined,
   type: 'button',
-  variant: 'default',
+  variant: 'neutral',
 };
 
-const Button = ({
-  children,
-  icon,
-  isBlock,
-  isCompact,
-  isDisabled,
-  isReversed,
-  size,
-  text,
-  type,
-  variant,
-  ...customProps
-  }) => {
-  const attributes = Object.assign({}, customProps);
-  const buttonText = text ? <span className={styles.text}>{text}</span> : null;
-
-  attributes.className = cx([
-    'button',
-    variant,
-    { 'is-disabled': isDisabled },
-    size,
-    { block: isBlock },
-    { compact: isCompact },
-    attributes.className,
-  ]);
-
-  attributes.type = type;
-  attributes.disabled = isDisabled;
-  attributes.tabIndex = isDisabled ? '-1' : undefined;
-  attributes['aria-disabled'] = isDisabled;
-
-  const order = isReversed ?
-    [buttonText, icon, children] :
-    [icon, buttonText, children];
-
-  return React.createElement(attributes.href ? 'a' : 'button', attributes, ...order);
-};
+class Button extends React.Component {
+  constructor(props) {
+    super(props);
+    this.state = { active: false, focused: false, mouseWasClicked: false };
+    this.handleKeyDown = this.handleKeyDown.bind(this);
+    this.handleKeyUp = this.handleKeyUp.bind(this);
+    this.handleOnBlur = this.handleOnBlur.bind(this);
+  }
+
+  handleOnBlur(event) {
+    this.setState({ focused: false });
+    event.stopPropagation();
```

Comment:
generally [stopping propagation is bad](https://css-tricks.com/dangers-stopping-event-propagation/). Why do we need to do this?